src/main/kotlin/io/github/anki/anki/service/secure/UserDetailsService.kt

```diff
@@ -0,0 +1,21 @@
+package io.github.anki.anki.service.secure
+
+import io.github.anki.anki.repository.mongodb.UserRepository
+import io.github.anki.anki.repository.mongodb.document.MongoUser
+import io.github.anki.anki.service.exceptions.UserDoesNotExistException
+import io.github.anki.anki.service.model.mapper.toUser
+import org.springframework.security.core.userdetails.ReactiveUserDetailsService
+import org.springframework.security.core.userdetails.UserDetails
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+
+@Component
+class UserDetailsService(
+    private val userRepository: UserRepository,
+) : ReactiveUserDetailsService {
+
+    override fun findByUsername(userName: String): Mono<UserDetails> =
+        userRepository.findByUserName(userName)
+            .switchIfEmpty(Mono.error(UserDoesNotExistException.fromUserName(userName)))
```

Comment:
it should be lazy:

```suggestion
            .switchIfEmpty {
                Mono.error(UserDoesNotExistException.fromUserName(userName))
            }
```

So, we don't create redundant objects when the publisher produces a value

Reply:
it says that switchIfEmpty accepts only Mono, not function

Reply:
Ok, then you need https://github.com/reactor/reactor-kotlin-extensions

Look at the implementation of the extension:
```
fun <T> Mono<T>.switchIfEmpty(s: () -> Mono<T>): Mono<T> = this.switchIfEmpty(Mono.defer { s() })

```

It "defers" the value, so if you apply the function then the error will be created on demand